linux/devdoc/srw_lock_linux_requirements.md

```diff
@@ -0,0 +1,146 @@
+# `srw_lock_linux` requirements
+
+## Overview
+
+`srw_lock_linux` is an implementation of slim read-write for linux.
+
+## Design
+
+`srw_lock_linux` used for synchronize resources across threads. It contains `exclusive mode` for write lock and `shared mode` for read lock. Acquire APIs will apply the required type of lock to `SRWLOCK` and get blocked if it cannot. Try acquire APIs will fail if the equivalent acquire APIs get blocked.
+
+## Exposed API
+
+```c
+typedef struct SRW_LOCK_HANDLE_DATA_TAG* SRW_LOCK_HANDLE;
+
+#define SRW_LOCK_TRY_ACQUIRE_RESULT_VALUES \
+    SRW_LOCK_TRY_ACQUIRE_OK, \
+    SRW_LOCK_TRY_ACQUIRE_COULD_NOT_ACQUIRE, \
+    SRW_LOCK_TRY_ACQUIRE_INVALID_ARGS
+
+MU_DEFINE_ENUM(SRW_LOCK_TRY_ACQUIRE_RESULT, SRW_LOCK_TRY_ACQUIRE_RESULT_VALUES)
+
+MOCKABLE_FUNCTION(, SRW_LOCK_HANDLE, srw_lock_create, bool, do_statistics, const char*, lock_name);
+
+/*writer APIs*/
+MOCKABLE_FUNCTION(, void, srw_lock_acquire_exclusive, SRW_LOCK_HANDLE, handle);
+MOCKABLE_FUNCTION(, SRW_LOCK_TRY_ACQUIRE_RESULT, srw_lock_try_acquire_exclusive, SRW_LOCK_HANDLE, handle);
+MOCKABLE_FUNCTION(, void, srw_lock_release_exclusive, SRW_LOCK_HANDLE, handle);
+
+/*reader APIs*/
+MOCKABLE_FUNCTION(, void, srw_lock_acquire_shared, SRW_LOCK_HANDLE, handle);
+MOCKABLE_FUNCTION(, SRW_LOCK_TRY_ACQUIRE_RESULT, srw_lock_try_acquire_shared, SRW_LOCK_HANDLE, handle);
+MOCKABLE_FUNCTION(, void, srw_lock_release_shared, SRW_LOCK_HANDLE, handle);
+
+MOCKABLE_FUNCTION(, void, srw_lock_destroy, SRW_LOCK_HANDLE, handle);
+```
+
+### srw_lock_create
+```c
+MOCKABLE_FUNCTION(, SRW_LOCK_HANDLE, srw_lock_create, bool, do_statistics, const char*, lock_name);
+```
+
+`srw_lock_create` creates a new `SRW_LOCK_HANDLE`.
+
+**SRS_SRW_LOCK_LINUX_07_001: [** `srw_lock_create` shall allocate memory for `SRW_LOCK_HANDLE_DATA`. **]**
+
+**SRS_SRW_LOCK_LINUX_07_002: [** `srw_lock_create` shall copy the `lock_name`. **]**
+
+**SRS_SRW_LOCK_LINUX_07_003: [** `srw_lock_create` shall initialized the `SRWLOCK` by calling `pthread_rwlock_init`. **]**
```

Comment:
>SRWLOCK [](http://example.com/codeflow?start=74&length=7)

This is a type from Windows ([https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-acquiresrwlockexclusive](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-acquiresrwlockexclusive)).

Likely we mean `pthread_rwlock_t ` here